keccak256/src/circuit.rs

```diff
@@ -0,0 +1,253 @@
+                let column = meta.advice_column();
+                meta.enable_equality(column.into());
+                column
+            })
+            .collect_vec()
+            .try_into()
+            .unwrap();
+
+        // theta
+        let theta_config = ThetaConfig::configure(meta.selector(), meta, state);
+        // rho
+        let rho_config = {
+            let cols: [Column<Advice>; 7] = state[0..7].try_into().unwrap();
+            let adv = RhoAdvices::from(cols);
+            let axiliary = [state[8], state[9]];
+
+            let base13_to_9 = [
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+            ];
+            let special = [meta.fixed_column(), meta.fixed_column()];
+            RhoConfig::configure(
+                meta,
+                state,
+                &adv,
+                axiliary,
+                base13_to_9,
+                special,
+            )
+        };
+        // Pi
+        let pi_config = PiConfig::configure(meta.selector(), meta, state);
+        // xi
+        let xi_config = XiConfig::configure(meta.selector(), meta, state);
+
+        // Iotab9
+        // Generate advice and instance column for Round constants in base9
+        let round_ctant_b9 = meta.advice_column();
+        let round_constants_b9 = meta.instance_column();
+        let iota_b9_config = IotaB9Config::configure(
+            meta,
+            state,
+            round_ctant_b9,
+            round_constants_b9,
+        );
+        let not_mixing_b9_to_13 = StateConversion::configure(meta);
+
+
+       
+        let mixing_config = MixingConfig::configure(meta, state);
+        // in side mixing  let b9_to_13 = StateConversion::configure(meta);
+
+
+        KeccakFConfig {
+            theta_config,
+            rho_config,
+            pi_config,
+            xi_config,
+            iota_b9_config,
+            mixing_config,
+            state,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn assign_all(
+        &self,
+        region: &mut Region<'_, F>,
+        mut offset: usize,
+        state: [F; 25],
+        out_state: [F;25],
+        flag: bool,
+        next_mixing: Option<[F; ABSORB_NEXT_INPUTS]>,
+        absolute_row_b9: usize,
+        absolute_row_b13: usize,
+    ) -> Result<[F; 25], Error> {
+        // In case is needed
+        let mut state = state;
+
+        // First 23 rounds
+        for round in 0..PERMUTATION {
+            // State in base-13
+            // theta
+            state = {
+                // Apply theta outside circuit
+                let out_state = KeccakFArith::theta(&state_to_biguint(state));
+                let out_state = state_bigint_to_pallas(out_state);
+                // assignment
+                self.theta_config
+                    .assign_state(region, offset, state, out_state)?
+            };
+
+            offset += ThetaConfig::OFFSET;
```

Comment:
I propose we use layouter to assign new regions for each state transition so that we don't have to remember all the offsets required.

Take theta for example, instead of taking the region and offset as input, we can take the layouter as the input
```diff
- self.theta_config.assign_state(region, offset, state, out_state)?;
+ self.theta_config.assign_state(layouter, state, out_state)?;
```